examples/ssh/opkssh/cli.go

```diff
@@ -0,0 +1,308 @@
+package main
+
+import (
+	"crypto"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/fs"
+
+	"os"
+	"path/filepath"
+	"strings"
+
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/exp/slices"
+
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/openpubkey/openpubkey/examples/ssh/sshcert"
+	"github.com/openpubkey/openpubkey/parties"
+	"github.com/openpubkey/openpubkey/pktoken"
+	"github.com/openpubkey/openpubkey/util"
+)
+
+// This function is called by the SSH server as the AuthorizedKeysCommand:
+//
+// The following lines are added to /etc/ssh/sshd_config:
+//
+//	AuthorizedKeysCommand /etc/opk/opkssh ver %u %t %k
+//	AuthorizedPrincipalsCommandUser root
+//
+// The parameters specified in the config map the parameters send the function below. They are:
+// %u The username (requested principal) - userArg
+// %t The public key type - typArg - in this case a certificate being used as a public key
+// %k The base64-encoded public key for authentication - certB64Arg - the public key is also a certificate
+func AuthorizedKeysCommand(userArg string, typArg string, certB64Arg string, policyEnforcer PolicyCheck, op parties.OpenIdProvider) (string, error) {
+	cert, err := sshcert.NewSshCertFromBytes(typArg, certB64Arg)
+	if err != nil {
+		return "", err
+	}
+	if err := CheckCert(userArg, cert, policyEnforcer, op); err != nil {
+		return "", err
+	} else {
+		// sshd expects the public key in the cert, not the cert itself.
+		// This public key is key of the CA the signs the cert, in our
+		// setting there is no CA.
+		pubkeyBytes := ssh.MarshalAuthorizedKey(cert.SshCert.SignatureKey)
+		return "cert-authority " + string(pubkeyBytes), nil
+	}
+}
+
+func CheckCert(userDesired string, cert *sshcert.PktSshCert, policyEnforcer PolicyCheck, op parties.OpenIdProvider) error {
+	pkt, err := cert.VerifySshPktCert(op)
+	if err != nil {
+		return err
+	}
+	err = policyEnforcer(userDesired, pkt)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type PolicyCheck func(userDesired string, pkt *pktoken.PKToken) error
```

Comment:
That's neat!

Reply:
I played around with this and decided not to do it, because I like having the definition of PolicyCheck in a single place that I can update rather than needing to define it in two function calls.